### PR TITLE
submitting transposeLDS & directToLds features

### PR DESCRIPTION
**Pull request by ramjana**

### CODE REVIEW

File: Tensile/Common.py [hunk 1]
```diff
@@ -704,6 +717,14 @@
     "LdsPadA":                     [ -1, 0, 1, 2, 3, 4, 8],
     "LdsPadB":                     [ -1, 0, 1, 2, 3, 4, 8],
 
+    # Padding boundary for LDS. defines block-size for pad insertion. for every 'LdsBlockSizePerPad' bytes, LDS padding (pad value from LdsPad parameter)
+    # is added (readOffset aware of the pad and adjusts offset value based on this parameter value).good rule of thumb is LdsBlockSizePerPad >= unrollDepth * BPE
+    # optimized value is 128
+    "LdsBlockSizePerPad":          [-1, 64, 128, 256],
```
Comment (AlexBrownAMD) [inline Tensile/Common.py]: Does -1 mean no padding?

File: Tensile/Common.py [hunk 2]
```diff
@@ -704,6 +717,14 @@
...
+
+    #Transpose LDS format. Local store in Coalsced dimension , same as optimized global fetch dimension . applicable only in TLU=0 case for miSIMD(s)
+    "TransposeLDS":                [-1, 1, 0],
```
Comment (AlexBrownAMD) [inline Tensile/Common.py]: What does -1 mean?

File: Tensile/Code.py [hunk 3]
```diff
@@ -281,6 +281,57 @@ class LocalReadInst (Inst):
   def __init__(self,*args):
     Inst.__init__(self,*args)
 
+################################################################################
+# MFMA Instruction
+################################################################################
+class  MFMAInst (Inst):
```
Comment (AlexBrownAMD) [inline Tensile/Code.py]: Is this class being used anywhere? Didn't see it referenced in the changeset

File: Tensile/KernelWriterAssembly.py [hunk 4]
```diff
@@ -4191,6 +4249,9 @@ def graFinalOffsets(self, kernel, tP):
                              "compute offset diff (tileDim)")
               else:
                 tileStride   = kernel[tP["lsp"]] * (perp*tVW + sPara*tVS)
+                #TODO remove wave_offset start for global fetching  TransposeLDS
```
Comment (AlexBrownAMD) [inline Tensile/KernelWriterAssembly.py]: Does this TODO need to be fixed before this checkin?

File: Tensile/KernelWriterAssembly.py [hunk 5]
```diff
@@ -4665,20 +4736,111 @@ def lwaFirstOffset(self, kernel, tP):
...
+        # use  parameter LdsBlockSizePerPAd parameter for adding pad size for lds  fetch efficiency
+        # calculate number of block chunks  and add ldsPAd
+        if not kernel["DirectToLds%s"%tP["tensorChar"]] :
+          LdsPadCnt = (((kernel["MacroTile%s"%tP["tensorChar"]] * kernel["DepthU"])*tP["bpe"])//wavefronts)//kernel["LdsBlockSizePerPad"]
+          WaveOffset = (((kernel["MacroTile%s"%tP["tensorChar"]] * kernel["DepthU"])*tP["bpe"])//wavefronts)
+        else:
+          #DirectToLds each global fetch 256 bytes, there is no vector Offset address for Writing into LDS
+          #m0 + instOffset is writeOffset; padding only accurs at 256 boundary
+          LdsPadCnt = 1
+          WaveOffset = 256
```
Comment (AlexBrownAMD) [inline Tensile/KernelWriterAssembly.py]: Always 256? Should this be based on parameters?

File: Tensile/KernelWriterAssembly.py [hunk 6]
```diff
@@ -4901,24 +5072,144 @@ def lraFinalOffset(self, kernel, tP):
...
+          kStr += inst("v_and_b32", \
+              vgpr(tP["gpr"]["lro"]), \
+              hex(kernel["SubGroup0"]-1), \
+              vgpr("Serial"), \
+              "vectorStaticDiv: %s = %s %% %u" \
+              % (vgpr(tP["gpr"]["lro"]), vgpr("Serial"), (kernel["SubGroup0"]-1)))
+        else:
+          kStr += inst("v_and_b32", \
+              vgpr(tP["gpr"]["lro"]), \
+              hex((((kernel["MatrixInstB"]*kernel["MatrixInstM"])//kernel["SubGroup0"])*kernel["MatrixInstN"])-1), \
```
Comment (AlexBrownAMD) [inline Tensile/KernelWriterAssembly.py]: (M*N*B)//MIWG0
Maybe make a variable for this calculation? It is similar to one done in a number of other places. Grouping MNB is easier to read. Also, I think we've used MIWG for MI calculations?

File: Tensile/KernelWriterAssembly.py [hunk 7]
```diff
@@ -4901,24 +5072,144 @@ def lraFinalOffset(self, kernel, tP):
...
+        #kStr += inst("v_add_u32", \
+        #    vgpr(tmpVgpraddr), \
+        #    vgpr(tmpVgpr), \
+        #    vgpr(tmpVgpraddr), \
+        #    "")
+        #kStr += inst("v_lshlrev_b32", \
+        #    vgpr("LocalReadAddr%s"%tc), \
+        #    log2(tP["bpe"]), \
+        #    vgpr(tmpVgpraddr), \
+        #    "")
```
Comment (AlexBrownAMD) [inline Tensile/KernelWriterAssembly.py]: Do we need this commented code?

File: Tensile/KernelWriterAssembly.py [hunk 8]
```diff
@@ -4901,24 +5072,144 @@ def lraFinalOffset(self, kernel, tP):
...
+            if tP["grcv"]:
+              PerpName = tP["lvp"]
+            else:
+              # Fractional load use the more accurate lsc, multiply by VW later
+              PerpName = tP["lsp"]
+          else:
+            if tP["grcv"]:
+              PerpName = tP["lvc"]
+            else:
+              PerpName = tP["lsc"]
```
Comment (AlexBrownAMD) [inline Tensile/KernelWriterAssembly.py]: Looks the same as above. Maybe should be a parameter in tP?

File: Tensile/KernelWriterAssembly.py [hunk 9]
```diff
@@ -4932,44 +5223,46 @@ def lraFinalOffset(self, kernel, tP):
...
+
+      # TODO: generalize over different MIs
+      if kernel["MatrixInstruction"] and not kernel["ProblemType"]["DataType"].isHalf():
+        if tc == "B" and "MatrixInstK" in kernel and kernel["MatrixInstK"] > 1 and tP["bpe"] == 4:
+          kDiv = kernel["MatrixInstN"]
+          kStr += vectorStaticDivide(rReg, dividendReg, kDiv, tmpVgpr, tmpSgpr)
+          kStr += inst("v_and_b32", \
+              vgpr(rReg), \
+              # 2 = 64 / 32 or number of ks
+              hex(2-1), \
```
Comment (AlexBrownAMD) [inline Tensile/KernelWriterAssembly.py]: Always 2-1? k can be 4 or 16 for smaller instructions

File: Tensile/KernelWriterAssembly.py [hunk 10]
```diff
@@ -6953,8 +7266,23 @@ def calculateLdsWriteOffset(self, perp, para, sPerp, sPara, kernel, tP, localWri
...
-    lspaOffset = perp_masked * kernel[tP["lsp"]]
-    rem = 0
+    if not kernel["TransposeLDS"]:
+      lspaOffset = perp_masked * kernel[tP["lsp"]]
+      rem = 0
+    else:
+      # for transposeLDS case,  Consecutive LDS  location contains elements from coalescing dimension
+      # lspa / num_wavefronts * (unrollDepth * BPE) + (unrollDepth*BPE/(LdsBlockSizePerPad/bpe) * LDsPAD)
+      # perInstWriteSize for each grvw = lspa//4 * unrollD
+      #TODO re-check
```
Comment (AlexBrownAMD) [inline Tensile/KernelWriterAssembly.py]: TODO

File: Tensile/KernelWriterAssembly.py [hunk 11]
```diff
@@ -6986,8 +7317,13 @@ def calculateLdsWriteOffset(self, perp, para, sPerp, sPara, kernel, tP, localWri
       lspaOffset *= (kernel[tP["mt"]] + kernel["LdsPad%s"%tc])
       lspaOffset += rem * ldl + perp_rem
     else:
-      lscaOffset *= (kernel[tP["mt"]] + kernel["LdsPad%s"%tc])
-      lscaOffset += rem
+      if not kernel["TransposeLDS"]:
+        lscaOffset *= (kernel[tP["mt"]] + kernel["LdsPad%s"%tc])
+        lscaOffset += rem
+      else:
+ 	#TODO re-verify lscaOffset=0
```
Comment (AlexBrownAMD) [inline Tensile/KernelWriterAssembly.py]: TODO

File: Tensile/KernelWriterAssembly.py [hunk 12]
```diff
@@ -7345,12 +7706,31 @@ def localReadDo(self, kernel, bufferIdx, iui, epsi, tP):
...
+                #TransposeLDS do not support ds_read2 opcodes
+                assert(numOffsets<2)
+                if tc == "A":
+                  InstructionTileOuput = kernel["MIWG0"]
+                else:
+                  InstructionTileOuput = (globalParameters["WavefrontWidth"] // kernel["MIWG0"] ) * (kernel["MatrixInstN"] // kernel["InstSplit"])
+                blockOffset = (InstructionTileOuput) * kernel["DepthU"] * tP["bpe"]
+                ldsPadOffset = (blockOffset//kernel["LdsBlockSizePerPad"])*kernel["LdsPad%s"%tc]*tP["bpe"]
+                offset = (blockOffset + ldsPadOffset)*rIdx + uIdx*kernel["MatrixInstK"]*tP["bpe"] + tP["localReadSwapByteOffset"]
+                paramList.append(offset)
```
Comment (AlexBrownAMD) [inline Tensile/KernelWriterAssembly.py]: Looks the same as the block above. Can this be simplified?

File: Tensile/SolutionStructs.py [hunk 13]
```diff
@@ -2515,11 +2522,31 @@ def assignDerivedParameters(state):
...
+    if state["TransposeLDS"] == 1:
+      if not state["MatrixInstruction"]:
+        reject(state, "TransposeLds Supports only in MatrixInstruction=1")
+    if "MatrixInstruction" in state:
+      if state["TransposeLDS"] == 1:
+        if state["ProblemType"]["TLUA"] or state["ProblemType"]["TLUB"]:
+          reject(state, "TransposeLds requires TLU=0")
+    if state["TransposeLDS"] == 1:
+      if state["LdsBlockSizePerPad"] == -1:
+        state["LdsBlockSizePerPad"] = 256
```
Comment (AlexBrownAMD) [inline Tensile/SolutionStructs.py]: Always 256?

File: Tensile/Common.py (hunk 1 above)
Comment (ramjana) [reply Tensile/Common.py]: -1 ; design choose value

File: Tensile/KernelWriterAssembly.py (hunk 4 above)
Comment (ramjana) [reply Tensile/KernelWriterAssembly.py]: lets keep the TODO for now

File: Tensile/KernelWriterAssembly.py (hunk 5 above)
Comment (ramjana) [reply Tensile/KernelWriterAssembly.py]: for DTL; offset is always 256

File: Tensile/KernelWriterAssembly.py (hunk 11 above)
Comment (ramjana) [reply Tensile/KernelWriterAssembly.py]: lets keep TODO For now..

File: Tensile/KernelWriterAssembly.py (hunk 12 above)
Comment (ramjana) [reply Tensile/KernelWriterAssembly.py]: they are not same.. for 'B' output space depends upon how wg0,wg1 mapped ; works out differently for mfma instructions

File: Tensile/SolutionStructs.py (hunk 13 above)
Comment (ramjana) [reply Tensile/SolutionStructs.py]: not best optimized value. but value for adding padding..

File: Tensile/KernelWriterAssembly.py (hunk 9 above)
Comment (ramjana) [reply Tensile/KernelWriterAssembly.py]: for f32 , that code will break in 16x16.. not sure who coded that.. definitely requires change..

File: Tensile/Common.py (hunk 1 above)
Comment (jichangjichang) [reply Tensile/Common.py]: how to disable "LdsBlockSizePerPad" concept to use legacy per unroll padding?

Comment (ramjana) [reply Tensile/Common.py]: by default LdsBlockSizePerPad = -1;  when TransposeLds=1 LdsBlockSizePerPad set to 256..

**Comment by rosenrodt** [comment]
This PR is crucial to fp16 performance so I checked it myself with [this config yaml](https://gist.github.com/rosenrodt/53594799a641e19871a4c1d76f18dcdc) and found a few failures cases:

| Macro-tile | Thread-tile | Failure condition |
|-----------------| ------|  -------------------|
| 64x128 | 1x32 | DU=8 |
| 64x256 | 2x32 | DU=8 |
| 32x256 | 1x32 | DU=8, 16 |
| 128x128 | 2x32 | All passed |

With `InnerUnroll=2` it doesn't seem to work with all above cases though

**Comment by ramjana** [comment]
> You should _not_ delete .jenkins/* since you're requesting a merge to develop, which needs CI to run.

added .jenkins back

**Comment by jichangjichang** [comment]
@ramjana @zaliu 
After this PR was merged, there are 6 failed cases in rocblas-test. It is related to directToLds. Please be informed. Thanks.
./rocblas-test --gtest_filter=\
nightly/gemm.blas3/f32_r_NT_784_1024_512_1_784_1024_0_784:\
nightly/gemm.blas3/f32_r_NT_784_1024_512_1_784_1024_1_784:\
nightly/gemm.blas3/f32_r_NT_1_1200_1_1_1_1200_0_1:\
nightly/gemm.blas3/f32_r_NT_1_4608_1_1_1_4608_0_1:\
nightly/gemm_ex.blas3/f32_rf32_rf32_rf32_rf32_r_NT_784_1024_512_1_784_1024_0_784_784:\
nightly/gemm_ex.blas3/f32_rf32_rf32_rf32_rf32_r_NT_784_1024_512_1_784_1024_1_784_784

**Comment by zaliu** [comment]
Tensile PR#889 fixes these 6 failures, I've confirmed that last night.  Thanks!

**Pull request merged**